src/protocol/attribution/credit_capping.rs

```diff
@@ -0,0 +1,477 @@
+
+        Ok(current_contribution.clone())
+    }
+
+    async fn compute_compare_bits<F: Field>(
+        ctx: SemiHonestContext<'_, F>,
+        current_contribution: &Batch<Replicated<F>>,
+        cap: u32,
+    ) -> Result<Batch<Replicated<F>>, Error> {
+        let cap = local_secret_shared_bits(cap.into(), ctx.role());
+        let one = Replicated::one(ctx.role());
+        let compare_bits: Batch<_> = try_join_all(
+            current_contribution
+                .iter()
+                .zip(zip(repeat(ctx.clone()), zip(repeat(cap), repeat(one))))
+                .enumerate()
+                .map(|(i, (contrib, (ctx, (cap, one))))| async move {
+                    let contrib_b = BitDecomposition::execute(
+                        ctx.narrow(&Step::BitDecomposeCurrentContribution),
+                        RecordId::from(i),
+                        contrib,
+                    )
+                    .await?;
+                    let lt_b = one.clone()
+                        - &BitwiseLessThan::execute(
+                            ctx.narrow(&Step::IsCapLessThanCurrentContribution),
+                            RecordId::from(i),
+                            &cap,
+                            &contrib_b,
+                        )
+                        .await?;
+                    Ok::<_, Error>(lt_b)
+                }),
+        )
+        .await?
+        .try_into()
+        .unwrap();
+        Ok(compare_bits)
+    }
+
+    async fn compute_final_credits<F: Field>(
+        ctx: SemiHonestContext<'_, F>,
+        input: &Batch<CreditCappingInputRow<F>>,
+        current_contribution: &Batch<Replicated<F>>,
+        compare_bits: &Batch<Replicated<F>>,
+        final_credits: &mut Batch<Replicated<F>>,
+        cap: u32,
+    ) -> Result<Batch<Replicated<F>>, Error> {
+        #[allow(clippy::cast_possible_truncation)]
+        let num_rows = input.len() as RecordIndex;
+        let cap = Replicated::from_scalar(ctx.role(), F::from(cap.into()));
+
+        /*
+        d = (cap - helperbits.get_vector(base = 1, size = numrows -1 )
+                * (cap + compare_bit.get_vector(base = 1, size = numrows -1)
+                    * ((cap - current_contribution.get_vector(base = 1, size = numrows-1)).get_vector())
+                    )
+                )
+        final_credits.assign_vector(
+            d.get_vector() + compare_bit.get_vector(base = 1, size = numrows-1)
+                * (final_credits.get_vector(base = 1, size = numrows - 1) - d.get_vector())
+            , base = 1)
+        */
+        // In the above logic, all vectors have "base = 1". Why are we skipping final_credits[0]??
+
+        for i in 0..(num_rows - 1) {
+            let a = cap.clone() - &current_contribution[i + 1].clone();
+            let b = cap.clone()
+                + &ctx
+                    .narrow(&Step::FinalCreditsSourceContribution)
+                    .multiply(RecordId::from(i), &compare_bits[i + 1].clone(), &a)
```

Comment:
I think that if you switch the order of arguments you can avoid a clone:
```
    .multiply(RecordId::from(i), a, &compare_bits[i+1])
```

Reply:
Fixed. Thanks!